fv/latency_test.go

```diff
@@ -0,0 +1,230 @@
+		}
+		etcd.Stop()
+	})
+
+	describeLatencyTests := func(c latencyConfig) {
+		var (
+			w   [2]*workload.Workload
+			cc  *workload.ConnectivityChecker
+			pol *api.GlobalNetworkPolicy
+		)
+
+		createPolicy := func(policy *api.GlobalNetworkPolicy) *api.GlobalNetworkPolicy {
+			log.WithField("policy", dumpResource(policy)).Info("Creating policy")
+			policy, err := client.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+			return policy
+		}
+
+		updatePolicy := func(policy *api.GlobalNetworkPolicy) *api.GlobalNetworkPolicy {
+			log.WithField("policy", dumpResource(policy)).Info("Updating policy")
+			policy, err := client.GlobalNetworkPolicies().Update(utils.Ctx, policy, utils.NoOptions)
+			Expect(err).NotTo(HaveOccurred())
+			return policy
+		}
+
+		BeforeEach(func() {
+			for ii := range w {
+				iiStr := strconv.Itoa(ii)
+				var ports string
+
+				ports = "3000"
+				w[ii] = workload.Run(
+					felix,
+					"w"+iiStr,
+					"cali0"+iiStr,
+					c.workloadIP(ii),
+					ports,
+					"tcp",
+				)
+
+				w[ii].DefaultPort = "3000"
+				w[ii].Configure(client)
+			}
+
+			cc = &workload.ConnectivityChecker{
+				Protocol: "tcp",
+			}
+
+			pol = api.NewGlobalNetworkPolicy()
+			pol.Namespace = "fv"
+			pol.Name = "policy-1"
+			pol.Spec.Ingress = []api.Rule{
+				{
+					Action: "Allow",
+				},
+			}
+			pol.Spec.Egress = []api.Rule{
+				{
+					Action: "Allow",
+				},
+			}
+			pol.Spec.Selector = "all()"
+
+			pol = createPolicy(pol)
+
+			cc.ExpectSome(w[0], w[1])
+			cc.ExpectSome(w[1], w[0])
+			cc.CheckConnectivity()
+		})
+
+		It("with allow-all should have good latency", func() {
+			meanRtt := w[0].LatencyTo(w[1].IP, w[1].DefaultPort)
+			_, err := fmt.Fprintf(resultsFile, "allow-all: %v\n", meanRtt)
+			Expect(meanRtt).To(BeNumerically("<", 10*time.Millisecond))
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Describe("with all() selector", func() {
+			BeforeEach(func() {
+				pol.Spec.Ingress[0].Source.Selector = "all()"
+				pol = updatePolicy(pol)
+			})
+
+			It("should have good latency", func() {
+				meanRtt := w[0].LatencyTo(w[1].IP, w[1].DefaultPort)
+				_, err := fmt.Fprintf(resultsFile, "all-selector: %v\n", meanRtt)
+				Expect(meanRtt).To(BeNumerically("<", 10*time.Millisecond))
+				Expect(err).NotTo(HaveOccurred())
+			})
+
+			Describe("with 10k IPs", func() {
+				BeforeEach(func() {
+					w[1].WorkloadEndpoint.Spec.IPNetworks = append(w[1].WorkloadEndpoint.Spec.IPNetworks,
+						generateIPs(10000)...)
+					wep := w[1].WorkloadEndpoint
+					wep.Namespace = "fv"
+					_, err := client.WorkloadEndpoints().Update(utils.Ctx, wep, utils.NoOptions)
+					Expect(err).NotTo(HaveOccurred())
+
+					Eventually(func() int {
+						return getNumIPSetMembers(felix.Container, "cali4-s:5y5I3VdRZfDU01O--xXAPx2")
+					}, "100s", "1000ms").Should(Equal(10002))
+				})
+
+				It("should have good latency", func() {
```

Comment:
Perhaps worth commenting here that the `all()` source selector will now expand to an IP set with 10000+ entries.